portfolio/src/main/java/com/google/sps/servlets/DataServlet.java

```diff
@@ -26,40 +32,52 @@
 @WebServlet("/data")
 public class DataServlet extends HttpServlet {
 
-  private List<String> messages;
   private static final String COMMENT = "comment-input";
   private static final String NAME = "name-input";
-
-  @Override
-  public void init() {
-    messages = new ArrayList<>();
-  }
+  private static final DatastoreService datastore = DatastoreServiceFactory.getDatastoreService();
+  private static final int MAX_RESULTS = 10;
+  private static final String COMM_TYPE = "Comment";
+  private static final String COMM_CONTENT = "content";
+  private static final String COMM_NAME = "name";
 
   @Override
   public void doPost(HttpServletRequest request, HttpServletResponse response) throws IOException {
-    // Construct comment from user request.
-    String fullComment = request.getParameter(NAME) + ": " + request.getParameter(COMMENT);
-    messages.add(fullComment);
+    String name = request.getParameter(NAME);
+    String comment = request.getParameter(COMMENT);
 
     response.setContentType("text/html;");
-    response.getWriter().println(fullComment);
+    response.getWriter().println(name + ": " + comment);
+
+    // Store comment in Datastore.
+    Entity commEntity = new Entity(COMM_TYPE);
+    commEntity.setProperty(COMM_CONTENT, comment);
+    commEntity.setProperty(COMM_NAME, name);
+    datastore.put(commEntity);
 
     // Redirect back to main page.
     response.sendRedirect("/index.html");
   }
 
   @Override
   public void doGet(HttpServletRequest request, HttpServletResponse response) throws IOException {
+    Query query = new Query(COMM_TYPE);
+    PreparedQuery results = datastore.prepare(query);
+    List<Entity> messages = results.asList(FetchOptions.Builder.withLimit(MAX_RESULTS));
+
     String json = convertToJson(messages);
     response.setContentType("application/json;");
     response.getWriter().println(json);
   }
 
-  private String convertToJson(List<String> messageList) {
+  private String convertToJson(List<Entity> messages) {
     String json = "{\"history\": [";
```

Comment:
I noticed in someone else's PR they are using Gson so convert this to a json string.

Reply:
yeah, using Gson would probably be more efficient/concise.